### DISCUSSION TITLE
Error when loading Tetgen tet10 meshes

### DISCUSSION THREAD

**Discussion by Edward-Eth** [libMesh/libmesh#3337] [General]
I have encountered a bug that I believe is occurring in libmesh (when used with MOOSE). When I load a tetgen tet10 mesh, the node orders are not read correctly leading to element self intersection:
<img width="639" alt="Screenshot 2022-07-14 at 14 51 29" src="https://user-images.githubusercontent.com/64844573/178998335-ca8161ce-4772-4b63-b502-b003d8174a3d.png">
I have made a tetgen file containing a single tet10 element, drawing the nodes out on paper shows that they're in the right places, and the nodes visible in paraview are in the correct place, it just seems that the order expected by libmesh and the order supplied by tetgen differ. 
The ele file as generated by tetgen:
<details>
<summary>ele file contents</summary>

```
6  10  0
    1       3     8     1     4      9    10    11    12    13    14
# Generated by tetgen -p ModelGeneration/cube.poly -o2 
```
</details>
The node file as generated by tetgen:
<details>
<summary>node file contents</summary>

```
27  3  0  0
   1    0  0  1
   3    1  1  1
   4    0  1  1
   8    0  1  0
   9    0  0.5  1
  10    0.5  1  1
  11    0.5  1  0.5
  12    0  0.5  0.5
  13    0  1  0.5
  14    0.5  0.5  1
# Generated by tetgen -p ModelGeneration/cube.poly -o2 
```
</details>
Not sure if this is any help, but this is the order the nodes are written in the tetgen file.
<img width="573" alt="Screenshot 2022-07-14 at 14 58 34" src="https://user-images.githubusercontent.com/64844573/178999996-e7847e04-e9f9-4854-9c80-fcd52e8d968c.png">
I have attempted a few "logical" feeling orders to rewrite these, but given there are several million permutations of 10 values I'm not holding out much hope for this scattershot approach...

**Comment by jwpeterson** [comment]
In tetgen_io.C, it looks like we just assume there is an identity mapping between the Tetgen numbering and our Tet10 numbering, but apparently that's not correct... 

https://github.com/libMesh/libmesh/blob/db05ce933ba4966a2a665a5ac2e1b9879b375bc1/src/mesh/tetgen_io.C#L209-L213 

Our Tet10 documentation shows the element numbering that we use:
https://github.com/libMesh/libmesh/blob/db05ce933ba4966a2a665a5ac2e1b9879b375bc1/include/geom/cell_tet10.h#L30-L63

and to me it looks like the main difference is that we number the mid-edge nodes between the {0,1,2} face of the Tet first whereas Tetgen numbers the mid-edge nodes in a different way that isn't immediately obvious to me. It should be possible to directly set up the correct `assign_elm_nodes` array given these two numberings.

**Reply by Edward-Eth** [reply]
Thanks for the quick response, I'll have a look at this and see if I can work out the mapping that's at work. If I figure it out I'll come back to you on it, would be good to fix this :)

**Reply by Edward-Eth** [reply]
Figured out the mapping:
If a tetgen element is described in the ele file using ten nodes:
`0 1 2 3 4 5 6 7 8 9`
then reordering them in the file to:
`0 1 2 3 6 7 9 5 8 4`
gives me the correct element. I'm not sure how that would translate to your assign_elm_nodes system, (maybe literally just like that?) but I hope that helps! In the mean time I'm just going to write a script to reorder each line of my tetgen ele file into that format and hopefully that should work.

**Reply by jwpeterson** [reply]
We just use `assign_elm_nodes` to map like this:
```
// Assign node to element
elem->set_node(assign_elm_nodes[j]) =
  mesh.node_ptr(/*libmesh global node id*/);
```
So based on (a 0-based version of) your screenshot above I would say
```
assign_elm_nodes = [0, 1, 2, 3, 9, 7, 4, 5, 8, 6]
```
Should do the trick. For example, this would map (0-based) Tetgen node 4 to libmesh node 9. To check this, observe that (0-based) Tetgen node 4 and libmesh node 9 are situated between (0-based) vertices {2,3} in both diagrams. 

Would it be possible for you to provide a non-reordered Tetgen input file that we could use as the basis of a unit test for this change in libmesh?

**Reply by Edward-Eth** [reply]
I redrew my diagram in the same orientation as the documentation diagram in cell_tet10.h and with zero indexing.
<img width="609" alt="Screenshot 2022-07-15 at 11 35 24" src="https://user-images.githubusercontent.com/64844573/179207078-51a653fa-72f2-483c-b9dc-a6a7f719115a.png">
Using this, and testing with a script to rewrite the order I have confirmed that the mapping is definitely:
`[0, 1, 2, 3, 6, 7, 9, 5, 8, 4]`
As for a unit test, you need the .ele file:
```
1  10  0
    1       1     2     3     4      5    6    7    8    9    10
# Generated by tetgen -p ModelGeneration/pyramid.poly -o2
```
and the .node file:
```
10  3  0  0
  1    1  1  1
  2    0  1  0
  3    0  0  1
  4    0  1  1
  5    0  0.5  1
  6    0.5  1  1
  7    0.5  1  0.5
  8    0  0.5  0.5
  9    0  1  0.5
  10    0.5  0.5  1
# Generated by tetgen -p ModelGeneration/pyramid.poly -o2 
```
this should create a pyramid with corners at (1,1,1) (0,0,1) (0,1,0) and (0,1,1) and a total volume of 0.166... Does that work as something you can implement? If the ordering is off then the volume would be.

**Reply by jwpeterson** [reply]
> Using this, and testing with a script to rewrite the order I have confirmed that the mapping is definitely:
> [0, 1, 2, 3, 6, 7, 9, 5, 8, 4]

Yes, this is the "inverse" to the `assign_elm_nodes` map that I posted in my previous comment, in other words,
```
inverse_map[assign_elm_nodes[j]] = j
```
for each `j`, so we are in agreement.

> Does that work as something you can implement?

Yes, I think so. I will hopefully make a PR some time this weekend that fixes the bug and adds a unit test. Thanks!

**Reply by jwpeterson** [reply]
#3338 was merged including the test case attached here.
